packages/itwinui-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -266,38 +275,51 @@ export const ComboBox = React.forwardRef(
         // Reset the filtered list (does not reset when multiple enabled)
         if (!multiple) {
           setFilteredOptions(optionsRef.current);
-          dispatch({ type: 'focus', value: undefined });
+          setFocusedIndex((selectedIndices as number | undefined) ?? -1);
         }
       }
       // When the dropdown closes
       else {
         // Reset the focused index
-        dispatch({ type: 'focus', value: undefined });
-        // Reset the input value if not multiple
-        if (!isMultipleEnabled(selected, multiple)) {
+        setFocusedIndex(-1);
+        // Reset/update the input value if not multiple
+        if (!isMultipleEnabled(selectedIndices, multiple)) {
           setInputValue(
-            selected != undefined && selected >= 0
-              ? optionsRef.current[selected]?.label
+            selectedIndices >= 0
+              ? optionsRef.current[selectedIndices]?.label
               : '',
           );
         }
       }
-    }, [isOpen, multiple, optionsRef, selected]);
+    }, [isOpen, multiple, optionsRef, selectedIndices]);
 
     // Update filtered options to the latest value options according to input value
     const [filteredOptions, setFilteredOptions] = React.useState(options);
+
+    // To reconfigure internals whenever the options change
     React.useEffect(() => {
-      if (inputValue) {
-        setFilteredOptions(
-          filterFunction?.(options, inputValue) ??
-            options.filter((option) =>
-              option.label.toLowerCase().includes(inputValue.toLowerCase()),
-            ),
+      // Remove the filter so that all of the new options are shown.
+      setFilteredOptions(options);
+
+      // If multiple=false, refocus the selected option.
+      // If no option is selected (i.e. selectedIndices === -1), reset the focus to the input.
+      if (!isMultipleEnabled(selectedIndices, multiple)) {
+        setFocusedIndex(selectedIndices as number);
+      }
+      // If multiple=true, reset the focus to the input.
+      else {
+        setFocusedIndex(-1);
+      }
+
+      // Reset/update the input value if multiple=false and if the dropdown is closed (i.e. don't override user input when dropdown is open)
+      if (!isMultipleEnabled(selectedIndices, multiple) && !isOpen) {
+        setInputValue(
+          selectedIndices >= 0
+            ? optionsRef.current[selectedIndices]?.label
+            : '',
         );
-      } else {
-        setFilteredOptions(options);
       }
-      dispatch({ type: 'focus', value: undefined });
+
       // Only need to call on options update
       // eslint-disable-next-line react-hooks/exhaustive-deps
     }, [options]);
```

Comment:
I'm a little confused by this particular `useEffect`. What do you mean by "reconfigure internals"?

Reply:
I was hoping to say something like "update the internal state (e.g. filteredOptions, focusedIndex, etc.) whenever the options change". I can change "internals" to state if you feel that's clearer. Or you can suggest a replacement too.

Unrelated: I will try to remove the `eslint-disable` by introducing a variable called something like `previousOptions` and run this logic whenever `options !== previousOptions`.

Reply:
> I will try to remove the `eslint-disable` by introducing a variable called something like `previousOptions` and run this logic whenever `options !== previousOptions`.

This is a good idea. I'll try to suggest improvements (if any) after you do this.

Reply:
> > I will try to remove the `eslint-disable` by introducing a variable called something like `previousOptions` and run this logic whenever `options !== previousOptions`.
> 
> This is a good idea. I'll try to suggest improvements (if any) after you do this.

Great, made this change (https://github.com/iTwin/iTwinUI/pull/2048/commits/f1a202da42e6309edb615767dce3e5bd3e829bad).

Reply:
If you're storing the previous state, then I think you could remove the `useEffect` altogether? ([docs](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes))

Reply:
Ohh, yes, I think it can be removed. Just wondering which of the two is better:
* not have a `useEffect` so that the rendering is essentially restarted to avoid rendering stale values
* using a `useLayoutEffect` instead so that the state is updated before the render

Reply:
`useLayoutEffect` still runs after React completes the first render. It is basically equivalent to `useEffect`, with the only difference being that it runs before the _browser_ paints. So there is no benefit to using `useLayoutEffect` over just changing the value during the render.

Reply:
Ohh, I see, sounds good. In [`dad30d0`](https://github.com/iTwin/iTwinUI/pull/2048/commits/dad30d0cac5691fc0669734a2ac476e94cc10d71), I removed the `useEffect` from here and also the `useLayoutEffect` from where `previousValue` was being used ([related conversation](https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1605156984)).

> I'm a little confused by this particular `useEffect`. What do you mean by "reconfigure internals"?

I renamed this to "reconfigure internal state" to try to make it clearer. Feel free to suggest a change if you had some other idea.

Reply:
Since you're reading the previous state during render now, you should store it in a state rather than ref. Refs are not supposed to be read or written during render. https://react.dev/learn/referencing-values-with-refs#differences-between-refs-and-state

Reply:
Upon testing after the changes from https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1605315949, I noticed many unexpected issues.

For example, in the below screen recording, if we select "AAA" and then change the `options` every 5 seconds:
* When the menu is open, the new options are not displayed automatically all the times. Manually re-clicking the ComboBox (something which I do in the second half of the recording) shows the latest `options`.
* Sometimes when the `options` changesm unselected options are shown as selected. I guess this is because 	`optionsExtraInfoRef` is not updating properly and so is returning incorrect/stale indexes for multiple options. So, doing `selected = _originalIndex === selectedIndices` returns `true` for multiple options.

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/bf562df3-478d-47e1-a6af-b378f5e639b2

</details>

Apart from these visual issues, I also came across some *runtime* errors in the browser that weren't highlighted in VSCode during *development* time. So, this could lead to unknown regressions if we don't test it extensively.

Most of these errors were "accessing variable before initialization" errors as a result of calling functions in the render function and not in the `useEffect`.

<details>
<summary>Some of the errors</summary>

<img width="792" alt="image" src="https://github.com/iTwin/iTwinUI/assets/45748283/0111add2-d270-41ac-afe0-faf1df786b44">

</details>

For now, I re-added back the `useLayoutEffect` since it works well and doesn't cause any issues (https://github.com/iTwin/iTwinUI/pull/2048/commits/7abdfd9a3d8f24a0cb94a35552e316c669cd173d). Also, from what I understand, `useLayoutEffect` updates the state and blocks browser painting until that state is updated ([Docs](https://react.dev/reference/react/useLayoutEffect#examples)). So, the user will not see any stale state. So, I think `useLayoutEffect` will be okay here.

<details>
<summary>Docs</summary>

<img width="889" alt="image" src="https://github.com/iTwin/iTwinUI/assets/45748283/e6e025be-89e6-4469-9032-f43e7ea8544e">

</details>

I tried some debugging with the `useEffect` and the `useLayoutEffect` removed (https://github.com/iTwin/iTwinUI/pull/2048/commits/cbf36bbb3d0d2b53535e5326b29dd0e39fd3573e). But since it's not related to what this PR is trying to solve, we can continue this later, if we indeed want to remove the `useLayoutEffect`s, since the docs says to avoid `useLayoutEffect` whenever possible.

Reply:
Did you try the previous state pattern mentioned [above](https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1605326112)? The issues you were facing might be because you were using ref instead of state.

Reply:
I tried using the "previous state" pattern but still couldn't get it to work without it giving any errors. My trial code is in [`8eb779e`](https://github.com/iTwin/iTwinUI/pull/2048/commits/8eb779e8f0858ab90298ad70dd2f651ac60ab5de). Here is what my trial includes:
* Move the `onOptionsChange()` and `onValuePropChange()` calls at the bottom of the render function to prevent the "variable used before initialization" runtime error. Doing this worked and didn't give any such runtime errors anymore, at least in the little that I tested it.
* Created a `#region` with explanatory comments to always keep the `onOptionsChange()` and `onValuePropChange()` calls at the bottom of the render function and not to move it elsewhere accidentally.
* I tested the case where the `options` passed to the ComboBox changes such that some of the options in the new `options` were not there in the old `options`. I constantly kept getting an error because `optionsExtraInfoRef` was updating too soon. i.e. `getMenuItem` was trying to render the old `options` but `optionsExtraInfoRef` now represented the new options.
  * To solve this error, one of the things I tried was moving around the logic that populates `optionsExtraInfoRef`. This didn't work.
  * To reproduce this error, you can `git checkout` the commit, open the vite playground, and then wait for a few seconds when the `options` changes automatically in `App.tsx`.

<details>
<summary>Error screenshot</summary>

<img width="815" alt="image" src="https://github.com/iTwin/iTwinUI/assets/45748283/541541cf-f4cc-4857-bbce-e5e64d44ce1f">

</details>

To solve the above-mentioned error when new `options` is passed, I thought of changing `optionsExtraInfoRef` from a ref to a state (https://github.com/iTwin/iTwinUI/pull/2048/commits/e18672648d1caf948a849e245775730319eda569). However, this too gave the same error. This commit also has the same repro steps as the ones mentioned above.